addons/mail/static/src/models/thread/thread.js

```diff
@@ -788,7 +788,14 @@ function factory(dependencies) {
          */
         _computeAllAttachments() {
             const allAttachments = [...new Set(this.originThreadAttachments.concat(this.attachments))]
-                .sort((a1, a2) => a1.id < a2.id ? 1 : -1);
+                  .sort((a1, a2) => {
+                      // Handle the temporary attachments that should be at the
+                      // start of the list. Temporary attachments have a negative id.
+                      if (a1.id > 0 && a2.id < 0) {
+                          return 1;
+                      }
+                      return a1.id < a2.id ? 1 : -1;
```

Comment:
Looks much better after this PR.

Just maybe the condition here should be inverted when both ids are below 0?
If the newest attachment should be displayed first, the one with -2 should be before the one with -1, just the inverse of position numbers.

Reply:
So you would have to handle 4 cases: both < 0, both > 0, a < 0 and b > 0, a > 0 and b < 0